katas/content/teleportation/Common.qs

```diff
@@ -1,7 +1,123 @@
+    // Helper which prepares proper Bell state on two qubits
+    // 0 - |Φ⁺⟩ = (|00⟩ + |11⟩) / sqrt(2)
+    // 1 - |Φ⁻⟩ = (|00⟩ - |11⟩) / sqrt(2)
+    // 2 - |Ψ⁺⟩ = (|01⟩ + |10⟩) / sqrt(2)
+    // 3 - |Ψ⁻⟩ = (|01⟩ - |10⟩) / sqrt(2)
+    operation StatePrep_BellState(q1 : Qubit, q2 : Qubit, state : Int) : Unit {
+        H(q1);
+        CNOT(q1, q2);
+        
+        // now we have |00⟩ + |11⟩ - modify it based on state arg
+        if state % 2 == 1 {
+            // negative phase
+            Z(q2);
+        }
+        
+        if state / 2 == 1 {
+            X(q2);
+        }
+    }
+
+    // ------------------------------------------------------
+    // Helper operation that run teleportation using the given operations to prepare the message qubit
+    // and the entangled pair, and to run sender and receiver parts of the protocol.
+    operation ComposeTeleportation(
+        bellPrepOp : ((Qubit, Qubit) => Unit), 
+        getDescriptionOp : ((Qubit, Qubit) => (Bool, Bool)), 
+        reconstructOp : ((Qubit, (Bool, Bool)) => Unit),
+        qAlice : Qubit, 
+        qBob : Qubit, 
+        qMessage : Qubit) : Unit {
+
+        bellPrepOp(qAlice, qBob);
+        let classicalBits = getDescriptionOp(qAlice, qMessage);
+        
+        // Alice sends the classical bits to Bob.
+        // Bob uses these bits to transform his part of the entangled pair into the message.
+        reconstructOp(qBob, classicalBits);
+    }
+    
+    operation SendMessage_Reference(qAlice: Qubit, qMessage: Qubit) : (Bool, Bool) {
+        CNOT(qMessage, qAlice);
+        H(qMessage);
+        return (M(qMessage) == One, M(qAlice) == One);
+    }
+
+    operation ReconstructMessage_Reference(qBob : Qubit, (b1 : Bool, b2 : Bool)) : Unit {
+        if b1 {
+            Z(qBob);
+        }
+        if b2 {
+            X(qBob);
+        }
+    }
+
+    // ------------------------------------------------------
+    // Helper operation that runs a teleportation operation (specified by teleportOp).
+    // The state to teleport is set up using an operation (specified by setupPsiOp).
+    //
+    // Specifying the state to teleport through an operation allows to get the inverse
+    // which makes testing easier.
+    operation TeleportTestHelper(
+        teleportOp : ((Qubit, Qubit, Qubit) => Unit), 
+        setupPsiOp : (Qubit => Unit is Adj)) : Bool {
+        
+        use (qMessage, qAlice, qBob) = (Qubit(), Qubit(), Qubit());
+        setupPsiOp(qMessage);
+            
+        // This should modify qBob to be identical to the state
+        // of qMessage before the function call.
+        teleportOp(qAlice, qBob, qMessage);
+            
+        // Applying the inverse of the setup operation to qBob
+        // should make it Zero.
+        Adjoint setupPsiOp(qBob);
+        if not CheckZero(qBob){
+            Message("Incorrect. The state was teleported incorrectly.");
+            DumpMachine();
```

Comment:
We want to print the state of Bob's qubit only, and we want to do it without Adjoint applied, so here we need to apply setupPsiOp(qBob) again (to get the state before adjoint application) and use DumpRegister([qBob]) to isolate one qubit from the rest

Reply:
Added setupPsiOp(qBob) and DumpRegister([qBob]) but this fails when qAlice and qBob qubits are entangled. So, called reset on qMessage and qAlice first and then followed up with the suggested steps.

Reply:
You're right, I noticed this when testing Bob's task, so I was thinking about reporting for the scenario when Alice did her part already, but if she didn't measure, the qubits can still be entangled... In this case measuring is not ideal for reporting on Alice's task; I got "state |0> was incorrectly teleported as |0>" on the first run, which is going to be super confusing... We should keep DumpMachine but add better message to explain the qubit order; let me leave a suggestion on that